File: Semestrul 4/MPP/lab03/java/src/main/java/mpp/repository/impl/EventRepositoryImpl.java
package mpp.repository.impl;

import mpp.domain.Event;
import mpp.repository.abstracts.AbstractDBRepository;
import mpp.repository.interfaces.EventRepository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class EventRepositoryImpl extends AbstractDBRepository<Long, Event> implements EventRepository {
    public EventRepositoryImpl() {
        super("Events", "id");
    }

    @Override
    protected String getInsertParameters() {
        return "name, date, location, description";
    }

    @Override
    protected String getUpdateParameters() {
        return "name=?, date=?, location=?, description=?";
    }

    @Override
    protected int getUpdateParameterCount() {
        return 4;
    }

    @Override
    protected Event createEntity(ResultSet data) throws SQLException {
        Long id = data.getLong("id");
        String name = data.getString("name");
        String date = data.getString("date");
        String location = data.getString("location");
        String description = data.getString("description");

        Event event = new Event(null, name, date, location, description);
        event.setId(id);
        return event;
    }

    @Override
    protected void saveEntity(PreparedStatement preparedStatement, Event entity) throws SQLException {
        preparedStatement.setString(1, entity.getName());
        preparedStatement.setString(2, entity.getDate());
        preparedStatement.setString(3, entity.getLocation());
        preparedStatement.setString(4, entity.getDescription());
    }

    @Override
    public List<Event> findByRefereeId(Integer refereeId) {
        String sql = "SELECT e.* FROM Events e " +
                "JOIN Results r ON e.id = r.event_id " +
                "WHERE r.referee_id = ?";

        List<Event> events = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setInt(1, refereeId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    events.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
//            logger.error("Error finding events by referee ID: " + e.getMessage());
        }

        return events;
    }
}

File: Semestrul 4/SDI/lab06/cs/Repository/Impl/EventRepositoryImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using cs.Domain;
using cs.Repository.Abstracts;
using cs.Repository.Interfaces;

namespace cs.Repository.Impl;

public class EventRepositoryImpl : AbstractDBRepository<long, Event>, IEventRepository
{
    public EventRepositoryImpl() : base("Events", "id")
    {
    }

    protected override string GetInsertParameters()
    {
        return "name, date, location, description";
    }

    protected override string GetUpdateParameters()
    {
        return "name=@name, date=@date, location=@location, description=@description";
    }

    protected override int GetUpdateParameterCount()
    {
        return 4;
    }

    protected override Event CreateEntity(IDataReader reader)
    {
        long id = Convert.ToInt64(reader["id"]);
        string name = reader["name"].ToString();
        string date = reader["date"].ToString();
        string location = reader["location"].ToString();
        string description = reader["description"].ToString();

        Event event_ = new(-1, name, date, location, description);
        event_.SetId(id);
        return event_;
    }

    protected override void SetParameters(SQLiteCommand command, Event entity)
    {
        command.Parameters.AddWithValue("@name", entity.GetName());
        command.Parameters.AddWithValue("@date", entity.GetDate());
        command.Parameters.AddWithValue("@location", entity.GetLocation());
        command.Parameters.AddWithValue("@description", entity.GetDescription());
    }

    public List<Event> FindByRefereeId(int refereeId)
    {
        string sql = "SELECT e.* FROM Events e " +
            "JOIN Results r ON e.id = r.event_id " +
            "WHERE r.referee_id = @refereeId";

        List<Event> events = new();

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@refereeId", refereeId);

            using var reader = command.ExecuteReader();
            while (reader.Read())
            {
                events.Add(CreateEntity(reader));
            }
        }
        catch (Exception ex)
        {
            // Logger.LogError($"Error finding events by referee ID: {ex.Message}");
        }

        return events;
    }
}